src/iamclient/iamclient.cpp

```diff
@@ -5,331 +5,630 @@
 
-        if (!remoteIM.mChannel) {
-            remoteIM.mChannel
-                = grpc::CreateCustomChannel(remoteIM.mConfig.mURL, mCredentials, grpc::ChannelArguments());
-        }
+    aos::iam::certhandler::CertInfo certInfo;
+    aos::Error                      err = AOS_ERROR_WRAP(mProvisionManager->ApplyCert(certType, pemCert, certInfo));
 
-        return CertificateService::NewStub(remoteIM.mChannel);
+    return SendApplyCertResponse(nodeId, certType, certInfo.mCertURL, certInfo.mSerial, err);
+}
+
+aos::Error IAMClient::CheckCurrentNodeStatus(const std::initializer_list<aos::NodeStatus>& allowedStatuses)
+{
+    aos::NodeInfo nodeInfo;
+
+    auto err = mNodeInfoProvider->GetNodeInfo(nodeInfo);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
     }
 
-    return nullptr;
+    const bool isAllowed = std::any_of(allowedStatuses.begin(), allowedStatuses.end(),
+        [currentStatus = nodeInfo.mStatus](const aos::NodeStatus status) { return currentStatus == status; });
+
+    return !isAllowed ? AOS_ERROR_WRAP(aos::ErrorEnum::eWrongState) : aos::ErrorEnum::eNone;
 }
 
-ProvisioningServiceStubPtr IAMClient::CreateIAMProvisioningServiceStub(const std::string& nodeID)
+aos::Error IAMClient::ExecCommand(const std::string& cmdName, const std::vector<std::string>& cmdArgs)
 {
-    std::lock_guard lock(mMutex);
+    if (!cmdArgs.empty()) {
+        std::string                    output;
+        const std::vector<std::string> args {cmdArgs.begin() + 1, cmdArgs.end()};
 
-    if (const auto it = mRemoteIMs.find(nodeID); it != mRemoteIMs.cend()) {
-        auto& remoteIM = it->second;
+        auto err = ExecProcess(cmdArgs[0], args, output);
+        if (!err.IsNone()) {
+            LOG_ERR() << cmdName.c_str() << " exec failed: output = " << output.c_str()
+                      << ", err = " << AOS_ERROR_WRAP(err);
 
-        if (!remoteIM.mChannel) {
-            remoteIM.mChannel
-                = grpc::CreateCustomChannel(remoteIM.mConfig.mURL, mCredentials, grpc::ChannelArguments());
+            return err;
         }
-
-        return ProvisioningService::NewStub(remoteIM.mChannel);
     }
 
-    return nullptr;
+    return aos::ErrorEnum::eNone;
 }
 
-/***********************************************************************************************************************
- * Private
- **********************************************************************************************************************/
+aos::Error IAMClient::ExecProcess(const std::string& cmd, const std::vector<std::string>& args, std::string& output)
+{
+    Poco::Pipe            outPipe;
+    Poco::ProcessHandle   ph = Poco::Process::launch(cmd, args, nullptr, &outPipe, &outPipe);
+    Poco::PipeInputStream outStream(outPipe);
+
+    Poco::StreamCopier::copyToString(outStream, output);
+    Poco::trimRightInPlace(output);
+
+    int exitCode = ph.wait();
+
+    return exitCode == 0 ? aos::ErrorEnum::eNone : aos::ErrorEnum::eFailed;
```

Comment:
Just return `exitCode` it will be converted to errno.

Reply:
moved to iamServer